### PR TITLE
CRF entity extraction Component

### PR DESCRIPTION
**Pull request by JoeyFaulkner**
Uses pycrfsuite to do entity extraction using conditional random fields. Conditional random fields are undirected graphical models which predict the most likely sequence of labels for a sentence given the words in that sentence. The parameters of the model attach weights to certain feature-label pairs, for example a feature might be "word.istitle() == True", which would mean there was a higher probability that that word has label "name" than if "word.istitle() == False" was the feature. We can also include features about immediately preceding and following words, including their POS tags.

For a more in-depth intro, see:
http://www.inference.phy.cam.ac.uk/hmw26/papers/crf_intro.pdf

### CODE REVIEW

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    """
+    Classification report for a list of BIO-encoded sequences.
+    It computes token-level metrics and discards "O" labels.
+    
+    Note that it requires scikit-learn 0.15+ (or a version from github master)
+    to calculate averages properly!
+    """
+    from sklearn.preprocessing import LabelBinarizer
+    from itertools import chain
+    from sklearn.metrics import classification_report, confusion_matrix
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: do we need to import confusion matrix?

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        print(bio_classification_report(y_test, y_pred))
+        
+        from collections import Counter
+        info = self.ent_tagger.info()
+
+
+
+
+def bio_classification_report(y_true, y_pred):
+    # taken from https://github.com/scrapinghub/python-crfsuite/blob/master/examples/CoNLL%202002.ipynb
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: not sure what the right thing to do is here. rasa NLU is under apache, which I think might not let us directly use MIT licensed code - even with attribution. Will look into this.

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        self.ent_tagger.open(self.f.name)
+
+    def _test_model(self, df_test):
+
+        X_test = [self._sentence_to_features(sent) for sent in df_test]
+        y_test = [self._sentence_to_labels(sent) for sent in df_test]
+        y_pred = [self.ent_tagger.tag(xseq) for xseq in X_test]
+        print(bio_classification_report(y_test, y_pred))
+        
+        from collections import Counter
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: what does this import do?

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            labels.append(word[2])
+        return labels
+
+    def _from_json_to_crf(self, json_eg):
+            #takes the json examples and switches them to a format which crfsuite likes
+            doc = self.nlp(json_eg[0])
+            entity_offsets = json_eg[1]
+            gold = GoldParse(doc, entities=entity_offsets)
+            ents = map(lambda l: l[5], gold.orig_annot)
+            if self.BILOU_flag == False:
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: `if not self.BILOU_flag` is more in line with style

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        for word in sentence:
+            labels.append(word[2])
+        return labels
+
+    def _from_json_to_crf(self, json_eg):
+            #takes the json examples and switches them to a format which crfsuite likes
+            doc = self.nlp(json_eg[0])
+            entity_offsets = json_eg[1]
+            gold = GoldParse(doc, entities=entity_offsets)
+            ents = map(lambda l: l[5], gold.orig_annot)
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: try and use list comprehension instead of map if possible, for py2/3 compatibility

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        labels = []
+        for word in sentence:
+            labels.append(word[2])
+        return labels
+
+    def _from_json_to_crf(self, json_eg):
+            #takes the json examples and switches them to a format which crfsuite likes
+            doc = self.nlp(json_eg[0])
+            entity_offsets = json_eg[1]
+            gold = GoldParse(doc, entities=entity_offsets)
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: goldparse is used for POS tags only right?

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    output_provides = ["entities"]
+
+    function_dict = {'low': lambda doc: doc[0].lower(), 'title': lambda doc: unicode(doc[0].istitle()),
+                     'word3': lambda doc: doc[0][-3:], 'word2': lambda doc: doc[0][-2:],
+                     'pos': lambda doc: doc[1], 'pos2': lambda doc: doc[1][:2],
+                     'bias': lambda doc: u'bias', 'upper': lambda doc: unicode(doc[0].isupper()),
+                     'digit': lambda doc: unicode(doc[0].isdigit())}
+
+    def __init__(self, ent_tagger=None, nlp=spacy.load('en'), crf_features=None, BILOU_flag=True):
```
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: nlp needs to be moved to the train / test / process methods. (We need to avoid loading anything from spacy anywhere, otherwise we'll end up with multiple instances of it at some point :D)

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    def test(self, testing_data):
+        if testing_data.num_entity_examples > 0:
+            test_data = self._convert_examples(testing_data.entity_examples)
+            ent_types = [[ent["entity"] for ent in ex["entities"]] for ex in testing_data.entity_examples]
+            entity_types = list(set(sum(ent_types, [])))
+            dataset = [self._from_json_to_crf(q) for q in test_data]
+            self._test_model(dataset)
+
+    # def process(self):
```
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: Can you clarify what is unclear here?

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    def _convert_examples(self, entity_examples):
+        def convert_entity(ent):
+            return ent["start"], ent["end"], ent["entity"]
+
+        def convert_example(ex):
+            return ex["text"], [convert_entity(ent) for ent in ex["entities"]]
+
+        return [convert_example(ex) for ex in entity_examples]
+
+    # def extract_entities(self, doc, nlp):
```
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: Can you clarify what is unclear here?

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            # features to use for each word, for example, 'title' in array before will have the feature "is the preceeding word in title case?"
+            self.crf_features = [['low', 'title', 'upper', 'pos', 'pos2'],
+                                 ['bias', 'low', 'word3', 'word2', 'upper', 'title', 'digit', 'pos', 'pos2'],
+                                 ['low', 'title', 'upper', 'pos', 'pos2']]
+        else:
+            self.crf_features = crf_features
+        self.config = {'crf_features': self.crf_features, 'BILOU_flag': BILOU_flag}
+
+    def train(self, training_data):
+        # type: training_data -> None
```
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: Should be `# type: (TrainingData) -> None`

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 8 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: so if i'm not mistaken the current extraction methods keep a self.ner function which will take a spacy doc and tag the words in the doc object with their entities and the doc object will then hold the new tagged entities. I was just going to make sure that this was necessary before I find some hacky way to do it that way.

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 9 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: see https://github.com/golastmile/rasa_nlu/pull/245/files/9b2a0eaf0ca1820d94f599144f0a4a471e268329#r109289617

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 6 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: that and converting the (start, end, entity) format entity_offsets to labelled words

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 5 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: 👍 will do

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 4 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: 👍

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 3 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: used it before for some diagnostics, removed!

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 2 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: is that true of pycrfsuite itself too or are we good on that?

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 10 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: 👍

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 6 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: but that can be done without it if necessary

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 2 above)
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: since we don't distribute pycrfsuite we are all good there. It's only an issue with code that we make available

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 8 above)
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: No you don't need to add the entities to the spacy doc object. The only thing you need to do is return them as a json object in the same way spacy and mitie do it:
```python
return {
         "entities": self.extract_entities(text, tokens, mitie_feature_extractor)
       }
```

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 6 above)
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: I think it is ok to use goldparse for now, since we need the spacy dependency for the POS tagging anyway (at least till we have another library to do the tagging)

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 2 above)
Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: License wise this is not an issue, (we can include code from MIT and publish as apache) need to include license and copyright though.

Comment (tmbo) [inline rasa_nlu/extractors/crf_entity_extractor.py]: Since this is a usage example anyway, I'd do the following: Add a copyright notice and the reference to MIT here in this function and thats it.

**Comment by tmbo** [comment]
Also,
- [ ] add change to Changelog
- [ ] add component to documentation
- [ ] python 3 compatibility
- [ ] add tests

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 11]
```diff
@@ -96,7 +96,7 @@ def load(cls, model_dir, model_name):
         if model_dir and model_name:
             ent_tagger = pycrfsuite.Tagger()
             ent_tagger.open(os.path.join(model_dir, 'ner', model_name))
-            config = json.load(open(os.path.join(model_dir, 'ner', 'crf_config.json'),'r'))
+            config = json.load(open(os.path.join(model_dir, 'ner', 'crf_config.json'), 'r'))
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: I think we want to use `io.open` here for py2/3 reasons

File: rasa_nlu/extractors/crf_entity_extractor.py [hunk 12]
```diff
@@ -114,10 +114,11 @@ def persist(self, model_dir):
 
             entity_extractor_config_file = os.path.join(ner_dir, "crf_config.json")
             entity_extractor_file = os.path.join(ner_dir, "model.crfsuite")
+            config = {'crf_features': self.crf_features, 'BILOU_flag': self.BILOU_flag}
             with open(entity_extractor_config_file, 'w') as f:
```
Comment (amn41) [inline rasa_nlu/extractors/crf_entity_extractor.py]: same here, `io.open` preferred

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 11 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: 👍

File: rasa_nlu/extractors/crf_entity_extractor.py (hunk 12 above)
Comment (JoeyFaulkner) [inline rasa_nlu/extractors/crf_entity_extractor.py]: 👍
